csi/volumeutils.py

```diff
@@ -1009,13 +1014,140 @@ def reload_glusterfs(volume):
+                elif line.startswith("subvolumes "):
+                    element.subvolumes = line.split(" ", 1)[-1]
+                elif line.startswith("type "):
+                    element.type = line.split()[-1]
+                    # print("type")
+                    # print(line.split()[-1])
+        return Volfile(volfile, elements)
+
+    def update_options_by_type1(self, opts):
+        for element in self.elements:
+            if opts.get(element.type, None) is not None:
+                #print(opts.get(element.type))
+                element.options.update(opts[element.type])
+
+    def update_options_by_type2(self, opts):
+        opt = {}
+        for element in self.elements:
+            for types, values in opts.items():
+                #print(element.type, types, values)
+                if element.type in types:
+                    print("before update", element.name, element.type)
+                    print(element.options)
+                    print("\n")
+                    element.options.update(values)
+                    print("after update", element.name, element.type)
+                    print(element.options)
+                    print("\n")
+                    opt[element.type] = element.options
+        return opt
+
+
+    def save(self, volfile=None):
+        filepath = self.volfile
+        if volfile is not None:
+            filepath = volfile
+
+        logging.info(logf(
+            "volfile path",
+            filepath=filepath
+        ))
+        with open(filepath + ".tmp", "w") as volf:
+            for element in self.elements:
+                #print(element.type)
+                #print(element.options)
+                volf.write("volume {0}\n".format(element.name))
+                volf.write("    type {0}\n".format(element.type))
+                for name, value in element.options.items():
+                    volf.write("    option {0} {1}\n".format(name, value))
+                if element.subvolumes is not None:
+                    volf.write("    subvolumes {0}\n".format(element.subvolumes))
+                volf.write("end-volume\n\n")
+
+        os.rename(filepath + ".tmp", filepath)
+
+
+def get_storage_options_hash(storage_options_sorted):
+    return xxhash.xxh64_hexdigest(storage_options_sorted)
+
+
+def sort_storage_options(storage_options):
+
+    sorted_options_tuple = sorted(storage_options.items(), key = lambda kv: kv[1])
+    storage_options_sorted = dict(sorted_options_tuple)
+    return storage_options_sorted
+
+
+def mount_glusterfs(volume, mountpoint, storage_options=None, is_client=False):
     """Mount Glusterfs Volume"""
     if volume["type"] == "External":
         volname = volume['g_volname']
     else:
         volname = volume["name"]
 
+
+    generate_client_volfile(volname)
+    client_volfile_path = os.path.join(
+        VOLFILES_DIR,
+        "%s.client.vol" % volname
+    )
+    if storage_options is not None:
+
```

Comment:
Use temp file as base file so that original Volfile(with only default options) will remain untouched.

```
import shutil
import tempfile

tmp_volfile = tempfile.mktemp(suffix=".vol")
shutil.copyfile(client_volfile_path, tmp_volfile)

volfile = Volfile.parse(tmp_volfile)
```

Then after finding the hash,

```
os.rename(tmp_volfile, new_client_volfile_path)
```

Reply:
```root@kadalu-csi-nodeplugin-wxp9c:/kadalu/volfiles# ls
storage-pool-1.client.vol  storage-pool-1_d2435480caff1b10.client.vol
root@kadalu-csi-nodeplugin-wxp9c:/kadalu/volfiles# diff storage-pool-1.client.vol storage-pool-1_d2435480caff1b10.client.vol -wB
32c27
<     option data-self-heal on
---
>     option data-self-heal off
36c31
<     option entry-self-heal on
---
>     option entry-self-heal off
71c64
<     option open-behind on
---
>     option open-behind off
```